File: modules/drive-sense/__tests__/timebase.test.ts
/** @jest-environment node */
// Review I2: the time base and the second windows the native ports group samples with.
import {
  MAX_ROW_GAP_MS,
  TIMEBASE_MAX_SKEW_MS,
  pickFix,
  takeWindow,
  toEpochMs,
  windowStart,
} from '../src/extract/timebase';
import type { FixSample } from '../src/extract/types';

const T0 = 1_700_000_000_000;
const anchor = { epochMs: T0, clockMs: 5_000_000 };

describe('toEpochMs', () => {
  test('converts through the capture anchor', () => {
    expect(toEpochMs(5_000_250, anchor, T0 + 900)).toEqual({ t: T0 + 250, fellBack: false });
  });

  test('a batched sample arriving 1 s late keeps its sensor time', () => {
    expect(toEpochMs(5_000_100, anchor, T0 + 1_100)).toEqual({ t: T0 + 100, fellBack: false });
  });

  test(`more than ${TIMEBASE_MAX_SKEW_MS} ms from arrival falls back to arrival`, () => {
    // a device whose sensor clock is not elapsedRealtime: hours off
    expect(toEpochMs(9_000_000, anchor, T0 + 500)).toEqual({ t: T0 + 500, fellBack: true });
    expect(toEpochMs(5_000_000 - TIMEBASE_MAX_SKEW_MS - 1, anchor, T0)).toEqual({ t: T0, fellBack: true });
    expect(toEpochMs(5_000_000 + TIMEBASE_MAX_SKEW_MS, anchor, T0)).toEqual({
      t: T0 + TIMEBASE_MAX_SKEW_MS,
      fellBack: false,
    });
  });
});

describe('windows', () => {
  test('the first row, and the row after a stall, take (ts − 1000, ts]', () => {
    expect(windowStart(null, T0 + 1000)).toBe(T0);
    expect(windowStart(T0, T0 + MAX_ROW_GAP_MS + 1)).toBe(T0 + MAX_ROW_GAP_MS + 1 - 1000);
  });

  test('otherwise (previous ts, ts] — a late timer loses nothing, an early one counts nothing twice', () => {
    expect(windowStart(T0 + 1000, T0 + 2300)).toBe(T0 + 1000); // timer slipped 300 ms: 1.3 s window
    expect(windowStart(T0 + 2300, T0 + 3000)).toBe(T0 + 2300); // caught up: 0.7 s window
  });

  test('consecutive windows over a jittery timer cover every sample exactly once', () => {
    const samples = Array.from({ length: 200 }, (_, k) => ({ t: T0 + 20 + k * 40 }));
    const ticks = [T0 + 1000, T0 + 2300, T0 + 3000, T0 + 3900, T0 + 5100, T0 + 6000, T0 + 7000, T0 + 8020];
    let buffer = samples;
    let prev: number | null = null;
    const seen: number[] = [];
    for (const ts of ticks) {
      const w = takeWindow(buffer, windowStart(prev, ts), ts);
      expect(w.dropped).toBe(0);
      seen.push(...w.inWindow.map((s) => s.t));
      buffer = w.rest;
      prev = ts;
    }
    expect(seen).toEqual(samples.filter((s) => s.t <= T0 + 8020).map((s) => s.t));
  });

  test('takeWindow sorts, keeps later samples, drops ones from a closed window', () => {
    const w = takeWindow([{ t: 5 }, { t: 3 }, { t: 12 }, { t: 1 }, { t: 10 }], 2, 10);
    expect(w.inWindow.map((s) => s.t)).toEqual([3, 5, 10]);
    expect(w.rest.map((s) => s.t)).toEqual([12]);
    expect(w.dropped).toBe(1);
  });

  test('the fix is chosen by its own timestamp, not by arrival order', () => {
    const fix = (t: number): FixSample => ({ t, lat: 0, lng: 0, hAcc: 5, speed: 1, speedAcc: 1, course: 0, alt: 0 });
    // arrival order: the 900 ms fix arrived after the next window's 1100 ms one
    const arrived = [fix(T0 + 1100), fix(T0 + 900), fix(T0 + 400)];
    expect(pickFix(arrived, T0, T0 + 1000)?.t).toBe(T0 + 900);
    expect(pickFix(arrived, T0 + 1000, T0 + 2000)?.t).toBe(T0 + 1100);
    expect(pickFix(arrived, T0 + 2000, T0 + 3000)).toBeNull();
  });
});

File: modules/drive-sense/android/src/test/java/expo/modules/drivesense/TimeBaseTest.kt
package expo.modules.drivesense

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The Android time base across a wall-clock step (review N2N3 I2; README §7 "Time base"). Pure JVM:
 * nothing here touches `SystemClock` — every instant is passed in on the boot clock.
 */
class TimeBaseTest {
  // The capture starts: wall 1_700_000_000_000, boot clock 50 000 ms.
  private val anchor = ClockAnchor(epochMs = 1_700_000_000_000.0, clockMs = 50_000.0)

  /** What the old code used as arrival: the wall clock, stepped forward by an hour mid-drive. */
  private fun steppedWallClock(clockMs: Double): Double = anchor.epochMs + (clockMs - anchor.clockMs) + 3_600_000.0

  @Test
  fun `a wall-clock step cannot trip the fallback when arrival is on the anchored boot clock`() {
    // a sensor sample stamped at boot 60 000 ms, delivered in a batch 900 ms later
    val stampMs = 60_000.0
    val arrival = TimeBase.anchoredNow(anchor, stampMs + 900.0)
    val c = TimeBase.toEpochMs(stampMs, anchor, arrival)
    assertFalse(c.fellBack)
    assertEquals(1_700_000_010_000.0, c.t, 0.0)

    // the same sample against the stepped wall clock: the old behaviour, pushed an hour off
    val old = TimeBase.toEpochMs(stampMs, anchor, steppedWallClock(stampMs + 900.0))
    assertTrue(old.fellBack)
    assertEquals(1_700_000_010_900.0 + 3_600_000.0, old.t, 0.0)
  }

  @Test
  fun `after the step the sample and the fix still land in their row's window`() {
    // Row ts comes from the same anchored boot clock (CaptureService.nowEpochMs).
    val ts = Math.round(TimeBase.anchoredNow(anchor, 61_000.0))
    val prevTs = Math.round(TimeBase.anchoredNow(anchor, 60_000.0))
    val start = TimeBase.windowStart(prevTs, ts)
    val sampleT = TimeBase.toEpochMs(60_500.0, anchor, TimeBase.anchoredNow(anchor, 61_300.0)).t
    val fix = FixSample(
      t = TimeBase.toEpochMs(60_800.0, anchor, TimeBase.anchoredNow(anchor, 61_200.0)).t,
      lat = 47.6, lng = -122.3, hAcc = 5.0, speed = 10.0, speedAcc = 0.5, course = 90.0, alt = 10.0
    )
    val w = TimeBase.takeWindow(listOf(sampleT), start, ts) { it }
    assertEquals(listOf(sampleT), w.inWindow)
    assertEquals(0, w.dropped)
    assertEquals(fix, TimeBase.pickFix(listOf(fix), start, ts))
  }

  @Test
  fun `a sensor stamping on another clock still falls back to its arrival`() {
    // stamps on uptime instead of elapsedRealtime: 20 000 ms behind
    val arrival = TimeBase.anchoredNow(anchor, 60_900.0)
    val c = TimeBase.toEpochMs(40_000.0, anchor, arrival)
    assertTrue(c.fellBack)
    assertEquals(arrival, c.t, 0.0)
  }

  @Test
  fun `windows follow the previous ts, except after a gap over MAX_ROW_GAP_MS`() {
    assertEquals(1_000L, TimeBase.windowStart(1_000L, 2_000L))
    assertEquals(4_000L, TimeBase.windowStart(1_000L, 5_000L))
    assertEquals(4_000L, TimeBase.windowStart(null, 5_000L))
    assertNull(TimeBase.pickFix(emptyList(), 0L, 1L))
  }
}
